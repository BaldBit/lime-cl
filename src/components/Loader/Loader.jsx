import React from 'react';
import PropTypes from 'prop-types';

import limeLight from '../Theme/limeLight';

import LoaderContainer from './styled';

const Loader = (props) => {
  const { size, speed, thickness, color } = props;

  return <LoaderContainer size={size} speed={speed} thickness={thickness} color={color} />;
};

Loader.propTypes = {
  /**
   * @property {string} [size=regular] - Set the size of the Loader component
   */
  size: PropTypes.oneOf(['regular', 'big']),
  /**
   * @property {number} speed - Set the animation speed
   */
  speed: PropTypes.number,
  /**
   * @property {number} thickness - Set the thickness of the Loader component
   */
  thickness: PropTypes.number,
  /**
   * @property {string} color - Set the color of the spinning circle
   */
  color: PropTypes.string,
};

Loader.defaultProps = {
  size: 'regular',
  speed: limeLight.animation.medium,
  thickness: limeLight.loader.thickness,
  color: limeLight.colorLight,
};

export default Loader;
