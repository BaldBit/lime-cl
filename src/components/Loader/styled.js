import styled, { keyframes, css } from 'styled-components';
import { rgba } from 'polished';

import limeLight from '../Theme/limeLight';

const loaderAnimation = keyframes`
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`;

const LoaderContainer = styled.div`
  position: relative;
  text-indent: -9999em;

  ${({ thickness, color, speed }) => css`
    border-top: ${thickness}px solid ${color};
    border-right: ${thickness}px solid ${color};
    border-bottom: ${thickness}px solid ${color};
    border-left: ${thickness}px solid ${rgba(color, 0.1)};

    animation: ${loaderAnimation} ${speed}s infinite linear;
  `};

  transform: translateZ(0);

  &,
  &:after {
    border-radius: 50%;
    ${({ size }) => `
      width: ${limeLight.loader.width[size]};
      height: ${limeLight.loader.height[size]};
    `}
  }
`;

export default LoaderContainer;
