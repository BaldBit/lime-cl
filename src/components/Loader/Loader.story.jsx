import React from 'react';
import keys from 'lodash-es/keys';

import { storiesOf } from '@storybook/react';
import { withKnobs, number, select } from '@storybook/addon-knobs';

import Loader from './Loader';
import limeLight from '../Theme/limeLight';

const stories = storiesOf('Loader', module);

stories.addDecorator(withKnobs);

stories.add('Default', () => {
  const speed = select('Speed', keys(limeLight.animation), 'medium');

  return (
    <Loader
      size={select('Size', ['regular', 'big'], 'regular')}
      speed={limeLight.animation[speed]}
      thickness={number('Thickness', limeLight.loader.thickness)}
      color={select('Color', ['red', 'blue', 'green'], 'red')}
    />
  );
});
