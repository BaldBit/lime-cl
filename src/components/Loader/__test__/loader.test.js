import React from 'react';
import { shallow } from 'enzyme';

import Loader from '../Loader';

it('renders correctly', () => {
  const wrapper = shallow(<Loader />);
  expect(wrapper).toMatchSnapshot();
});
