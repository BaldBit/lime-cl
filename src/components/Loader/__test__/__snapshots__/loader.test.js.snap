// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<styled.div
  color="#ffffff"
  size="regular"
  speed={0.5}
  thickness={2}
/>
`;
